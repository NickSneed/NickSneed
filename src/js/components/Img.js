import { LazyLoadImage } from 'react-lazy-load-image-component';
import PropTypes from 'prop-types';

// Function to animate the loading of an image
const loadingAnimation = (e) => {
    setTimeout(function () {
        e.target.style.opacity = 1;
    }, 50);
}

// Img component
const Img = (props) => {

    const {
        src, 
        srcSmall, 
        alt, 
        className, 
        percent, 
        width, 
        height 
    } = props;
    const style = {
        opacity: 0,
        transition: 'opacity 1s linear 0s'
    }

    // Base HTML for an image
    let html = (
        <LazyLoadImage
            style={style}
            src={src}
            alt={alt}
            width={width}
            height={height}
            className={className}
            onLoad={function (e) {
                loadingAnimation(e)
            }} />
    )

    // If a small image is provided, wrap the image in a picture element
    if (srcSmall) {
        html = (
            <picture>
                <source srcSet={srcSmall} media="(max-width: 959px)" />
                {html}
            </picture>
        )
    }

    // If a percentage is provided, wrap the image in a div with a spacer
    if (percent) {
        html = (
            <div className={className ? 'img-percent ' + className : 'img-percent'}>
                <div className="spacer" style={{
                    'paddingTop': percent + '%'
                }}></div>
                {html}
            </div>
        )
    }

    return html;
}

// Define the prop types for the component
Img.propTypes = {
    src: PropTypes.string.isRequired,
    srcSmall: PropTypes.string.notRequired,
    alt: PropTypes.string.isRequired,
    className: PropTypes.string.notRequired,
    percent: PropTypes.string.notRequired,
    width: PropTypes.string.notRequired,
    height: PropTypes.string.notRequired
};

export default Img;