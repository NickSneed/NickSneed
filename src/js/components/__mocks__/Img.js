/* eslint-disable react/prop-types */
import React from 'react';

const Img = ({ src, alt, ...props }) => {
    return <img src={src} alt={alt} data-testid="mock-img" {...props} />
}

export default Img;
