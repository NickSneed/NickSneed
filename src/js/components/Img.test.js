/* global describe, test, expect, jest */
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
jest.unmock('@/js/components/Img.js');
import Img from '@/js/components/Img';

describe('Image Component', () => {
    const mockProps = {
        src: 'test-image.jpg',
        alt: 'Test image',
        className: 'test-class',
        percent: '100'
    };

    test('renders image with required props', () => {
        render(<Img src={mockProps.src} alt={mockProps.alt} />);
        const image = screen.getByAltText(mockProps.alt);
        expect(image).toBeInTheDocument();
        expect(image).toHaveAttribute('src', mockProps.src);
    });

    test('renders image with all props', () => {
        render(<Img {...mockProps} />);
        const image = screen.getByAltText(mockProps.alt);
        expect(image).toHaveClass(mockProps.className);
    });

    test('renders picture element when srcSmall is provided', () => {
        render(<Img {...mockProps} srcSmall="small-image.jpg" />);
        const picture = document.querySelector('picture');
        expect(picture).toBeInTheDocument();
        expect(picture.querySelector('source')).toHaveAttribute('srcset', 'small-image.jpg');
    });

    test('renders wrapper div when percent prop is provided', () => {
        render(<Img {...mockProps} />);
        const wrapper = document.querySelector('div > div');
        expect(wrapper).toHaveStyle({ position: 'relative', width: '100%' });
    });
});