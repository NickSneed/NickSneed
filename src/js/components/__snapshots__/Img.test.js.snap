// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Component matches snapshot 1`] = `
<div>
  <div
    class="css-0"
    style="position: relative; width: 100%;"
  >
    <div
      class="css-0"
      style="padding-top: 100%; height: 0px; width: 100%;"
    />
    <picture>
      <source
        media="(max-width: 959px)"
        srcset="small-image.jpg"
      />
      <img
        alt="Test image"
        class="test-class"
        src="test-image.jpg"
        style="display: block; height: 100%; width: 100%; opacity: 0; transition: opacity 1s linear 0s; max-width: none; margin: 0px auto; position: absolute; left: 0px; top: 0px; object-fit: contain;"
      />
    </picture>
  </div>
</div>
`;
